reladomographql/docs/temporal-milestoning.md

```diff
@@ -0,0 +1,81 @@
+# Temporal Milestoning
+```graphql
+query {
+  accounts (findMany: {processingDate: {eq: "2019-10-13T04:04:39+00:00"}})
+  {
+    accountNum
+    processingDateFrom
+    processingDateTo
+  }
+}
+
+```
+Here is result. Note how the processingDate input belongs to the interval of [processingDateFrom, processingDateTo) of the data.
+
+```json
+{
+  "data": {
+    "accounts": [
+      {
+        "accountNum": "100002",
+        "processingDateFrom": "\"2019-01-03T08:00:00Z\"",
+        "processingDateTo": "\"9999-12-02T04:59:00Z\""

```

Comment:
`null` is not a particularly good representation of infinity. Reladomo actually supports the null value as infinity, but it's highly discouraged, so returning null here is also confusing from that perspective. 

I suggest fixing the bug.

Reply:
I strongly agree with an infinity date like 9999 in the database. I still think it may make sense to translate between null/infinity during serialization/deserialization.

Server-side code has easy access to the infinity date (from `DefaultInfinityTimestamp.getDefaultInfinity()`, AsOfAttribute.getInfinityDate(), and possibly other places). Client-side code won't have this information as easily available.

If the service does return infinity dates inside json, then I think it also makes sense to add an endpoint that just serves the infinity date as a constant.

Reply:
What do you think about supporting the infinity as an explicit scalar. Something like "INFINITY". it can be part of the TimestampScalar. I will need to figure out how to pass it around java, but it is doable since everything is in the context of mithra attributes and mithra attributes have the concept of infinity.

more interesting question is what to do with timezones. The DB data is in the local timezone. Depending on where the JVM runs, the timezone is different and the GMT off-set if different. 
Does it mean that all timestamps should be in UTC in the test databases?

Reply:
Infinity is just a constant. That's not the only place a model will have constants, e.g. a synonym type of "CUS" means cusip. Unless you're going to solve the more general problem of constants, I'd say just use the value.

Timezones are properly represented in the database and the jvm because we take care of proper serialization/deserialization. You have to do something similar between the jvm and the js runtime.

Reply:
oh, and the serialization code deals with infinity explicitly.

Reply:
I've written REST POST/PUT methods that allow "infinity" in the place of dates. I've never written any GraphQL mutations, only queries. I think both "infinity" and `null` are nice conveniences.

I'll work on open-sourcing my GraphQL work so we can collaborate more easily. For now, I'll just include one snippet inline - the DataFetcher for timestamp attributes. Note the hoops I had to jump through to get the time to be in the right timezone. Just calling timestamp.toInstant() gave me the wrong result. Now that I think about it, I'm not sure if I'm supposed to use CONVERT_NONE, but I'm using CONVERT_TO_UTC.

```java
public class ReladomoInstantDataFetcher<Input> implements DataFetcher<Instant>
{
    private final TimestampAttribute<Input> timestampAttribute;

    public ReladomoInstantDataFetcher(TimestampAttribute<Input> timestampAttribute)
    {
        this.timestampAttribute = timestampAttribute;
    }

    @Nullable
    @Override
    public Instant get(@Nonnull DataFetchingEnvironment environment)
    {
        Input persistentInstance = environment.getSource();
        if (persistentInstance == null)
        {
            return null;
        }

        if (this.timestampAttribute.isAttributeNull(persistentInstance))
        {
            return null;
        }

        Timestamp     result        = this.timestampAttribute.valueOf(persistentInstance);
        LocalDateTime localDateTime = result.toLocalDateTime();
        return localDateTime.toInstant(ZoneOffset.UTC);
    }
}
```

Reply:
@motlin : I think we are on the same page on how to convert the UTC time zone. In my code, it is implemented as TimestampScalar (https://github.com/vborisoff/reladomo/blob/graphql-real/reladomographql/src/main/java/com/gs/reladomo/graphql/TimestampScalar.java)

The question is how to represent the infinity from a mathematical point of view.  Let say we have DB in NY. it stores infinity (out_z) as 23:59 EST and the insert time (in_z) as 15:00 EST. When I send it out I want to send it out in UTC. Both the infinity and the insert time have to be converted. The insert time formated as "20:00 UTC". The infinity is formated as "04:59 UTC". This is the mathematically correct answer: the infinity is indeed 04:59 UTC. it just looks ugly. To make it look prettier, we can format the timestamp in the database time zone. Then the infinity will be "23:59 EST". 

Alternatively we can make all timestamps in the UTC format in the database.

Reply:
I don't think that's a sensible way to think about infinity. Infinity doesn't have a timezone. This is already fully fleshed out in the Reladomo code and you should keep it consistent. Have you looked at how we serialize infinity? From `MithraTimestamp`:

```java
    public static void writeTimestampWithInfinity(DataOutput out, Timestamp timestamp, Timestamp infinity) throws IOException
    {
        if (timestamp == null)
        {
            out.writeByte(IS_NULL);
            return;
        }
        if (timestamp.getTime() == infinity.getTime())
        {
            out.writeByte(IS_INFINITY);
            return;
        }
        writeNormalTimestamp(out, timestamp.getTime(), timestamp.getNanos());
    }
```

We write a single byte. No date/time or anything else. Just a flag.

Reply:
@mohrezaei  makes all sense to me. The question is how to represent infinity in the string format in JSON. Other timestamps are printed in the ISO8601/UTC formant (aka RFC-3339).
 I can consider different approaches to print infinity:
- print it as "9999-12-01T23:59:00Z' - not the same as the database timzone, but the meaning is clear.
- print the word "INFINITY" or maybe "00" as an approximation of the infinity symbol.
- print character for 0x20 (space?) to be consistent with writeTimestampWithInfinity.

In any case, I need to have a specialized DataFetcher and associated attribute. I like Craig's approach for this and will try to implement it as the dynamically generated schema. I would be interested to catch up on this, Craig.

Reply:
I'm ok with "9999-12-01T23:59:00Z" or "INFINITY". JSON can do full utf-8, so you can use the real infinity symbol ∞ (U+221E), but I would advise against that. (0x20 is a private representation and should never be used in publicly visible contexts).

Reply:
Is it shown as 'INFINITY' in the result and can be used as 'INFINITY' in queries. 
The 'INFINITY' is matched to the asOfAttribute infinity or null in case if the asOfAttribute Infinity is null parameter is set.
Here are some examples: https://github.com/vborisoff/reladomo/blob/graphql-real/reladomographql/docs/temporal-milestoning.md